### ISSUE TITLE
Issue with copy/paste into py65mon on linux

### ISSUE THREAD

**Issue by SamCoVT** [mnaberez/py65#47]
I'm having an issue where input pasted into py65mon running in a linux terminal (tried xterm, konsole, and gome-term) gets "stuck" and new keys have to be pressed on the keyboard in order to get the pasted input to be processed one letter at a time.

I believe the issue is in utils/console.py where it switches in and out of non-canonical mode when trying to get a single character.  The issue occurs when there is already waiting input at stdin, and this situation is easy to create when pasting into the terminal window running py65mon.  The characters are entered much faster than py65mon can use them up, so the remaining characters are held in the input buffer and not read at that time.  To get the buffered characters to be read, new input must be typed in.  Typing new characters starts echoing and processing the old characters stuck in the buffer, and the new characters are lost.  Once all of the old characters have been processed, the new characters being typed start being processed, as if py65mon has "caught up" to the input.
### Reproducing
I believe this is reproducible with any 6502 program using the I/O routines built into the simulator.  Just have your 6502 program read and echo, then paste into the terminal window running p65mon.  Only the first character (or perhaps few characters) appears.  Then, type anything (I just hold down the spacebar) to get the rest of the pasted input to be processed.

I'm currently seeing this when working on Tali Forth 2 in the simulator.  It happens super-rarely when typing really fast, but always happens when I paste into the terminal window.  It doesn't matter if I use middle-click paste or Edit->paste from the terminal menus (those are two different clipboards.)  If you clone the Tali Forth 2 repository (https://github.com/scotws/TaliForth2.git) you can just `make sim` to get Tali2 running in the simulator.

### Some thoughts
I can paste into a minicom session attached to my real hardware (now that I have hardware handshaking working on my hardware) and that works fine, so I know it's not a problem in the 65C02 software I'm simulating in py65mon.

I believe the code is switching back and forth between non-canonical mode and canonical mode in order to be able to hand characters immediately to the software being run in the simulator, but still have line editing functions for the monitor.  Does it make sense to switch to canonical mode only for the monitor, and leave it in non-canonical mode the rest of the time?  

My reading of the C libraries seems to indicate that setting the terminal attributes with TCSANOW should actually discard any unprocessed input, but it doesn't seem to do that.  It's possible there is an extra input buffer in Python itself that is causing this weird behavior.

Python is not my first language, but I'll poke around and see if I can figure out a fix for my issue.

**Comment by SamCoVT** [comment]
I've made some progress, and I was able to paste approx 60K of input into the terminal running py65mon and have it all be passed to the 65C02 software running inside, but my fix has some side effects that need to be figured out.  

TLDR version - skip to the end for my suggested fix


I'm currently playing in utils/console.py in getch_noblock() and I have a few notes.  The code I am currently working with looks like this:
```
    def getch_noblock(stdin):
        """ Read one character from stdin without blocking.  Does not echo the
        character.  If no character is available, an empty string is returned.
        """

        fd = stdin.fileno()

        oldterm = termios.tcgetattr(fd)
        newattr = oldterm[:]
        newattr[3] = newattr[3] & ~termios.ICANON & ~termios.ECHO
        termios.tcsetattr(fd, termios.TCSANOW, newattr)

        oldflags = fcntl.fcntl(fd, fcntl.F_GETFL)
        fcntl.fcntl(fd, fcntl.F_SETFL, oldflags | os.O_NONBLOCK)

        try:
            char = ''
#            r, w, e = select.select([fd], [], [], 0.1)
#            if r:
            #print("!") # DEBUG
            # The next 3 lines were outdented one step
            char = stdin.read(1)
            if char == "\n":
                char = "\r"
        finally:
#            termios.tcsetattr(fd, termios.TCSAFLUSH, oldterm)
            termios.tcsetattr(fd, termios.TCSANOW, oldterm)
            fcntl.fcntl(fd, fcntl.F_SETFL, oldflags)
        return char
```
My first note is that if you are switching STDIN to be non-blocking, then you don't need the select.  Normally you would use select to wait on the file descriptor so that you would know when you could read on a _blocking_ file descriptor and be guaranteed not to block (at least for 1 character, anyway).  With a _non-blocking_ file descriptor, just go ahead and try to read.  If there is no data, it won't return any.  If there is data, it will give you the 1 character you are asking for.  

If you don't like the high CPU usage of doing a non-blocking read over and over until a character shows up (and I'm guessing this was an issue by the 0.1 timeout in the select statement), you can just put your own delay in, after the read, if you didn't get anything.

The second note is that there appears to be two input buffers on stdin.  I'm guessing that one is in the C library/OS and the other is in Python.  If I switch back to the commented line in the finally section that sets the termios attributes back they way they were (back to canonical mode) using the TCSAFLUSH option, when I paste, I only get about 4K or so and all the rest is lost.  The C reference for TCSAFLUSH says that all input so far received but not read is discarded.  I believe the C/OS input buffer is indeed being discarded, but the Python input buffer keeps it's data.  This also explains the original behavior, where the select() needs additional input to actually try to read, but the read gets the data that was already in the Python input buffer.

Some unresolved issues:
If you hit CTRL-C at just the right time to get back to the monitor, sometimes the input is not echoed and you are typing blind.  It doesn't do this most of the time, but I have seen it a couple of times now, and I suspect it's happening because the CTRL-C was processed after the echo was turned off, but before it was turned back on again.
Sometimes, when typing really fast, a character appears twice on the screen, but only one copy is given to the program.  I suspect this is the same problem but the other way around - the char comes in while echo is on (so Linux echoes it), but then my program reads it and echoes it as well.

### Proposed Fix
I propose turning echo off and switching to non-canonical mode while the simulator is "running", and turn echo back on and switch to canonical mode when the monitor prompt comes back up, rather than switching back and forth on a character by character basis.  I also suggest using TCSANOW on all calls that switch back and forth to prevent lost input.  I'll have to dig into the monitor to see where the best place to put this might be.  If you have any suggestions on where to start looking, that would help me a lot.

**Comment by SamCoVT** [comment]
It's worth noting that this is the same section of code being dealt with in #46, although this is a different issue.  I haven't seen #46 happen at all, but the error messages an ioctl so that makes me think they are related.

**Comment by mnaberez** [comment]
Fixed in 5ae7eeb6f79659f13a3b7e7a85eb80e7561aa984.
